django/utils/deprecation.py

```diff
@@ -83,6 +85,204 @@ def __new__(cls, name, bases, attrs):
+        num_by_kind = Counter(param.kind for param in params.values())
+
+        if num_by_kind[inspect.Parameter.VAR_POSITIONAL] > 0:
+            raise TypeError(
+                "@deprecate_posargs() cannot be used with variable positional `*args`."
+            )
+
+        num_positional_params = (
+            num_by_kind[inspect.Parameter.POSITIONAL_ONLY]
+            + num_by_kind[inspect.Parameter.POSITIONAL_OR_KEYWORD]
+        )
+        num_keyword_only_params = num_by_kind[inspect.Parameter.KEYWORD_ONLY]
+        if num_keyword_only_params < 1:
+            raise TypeError(
+                "@deprecate_posargs() requires at least one keyword-only parameter"
+                " (after a `*` entry in the parameters list)."
+            )
+        if any(
+            name not in params or params[name].kind != inspect.Parameter.KEYWORD_ONLY
+            for name in moved
+        ):
+            raise TypeError(
+                "@deprecate_posargs() `moved` names must"
+                " all be keyword-only parameters."
+            )
+
+        num_moveable_args = len(moved)
+        max_positional_args = num_positional_params + num_moveable_args
+
+        param_names = list(params.keys())
+
+        # The deprecation message logic needs to know if func is a bound method.
+        # Because decorators always receive plain, unbound functions, the class
+        # information is not available here. (inspect.ismethod(func) will always
+        # return False, and func.__self__ will never be set.) This only affects
+        # message construction, not argument mapping logic, so it's fine to just
+        # look for the customary bound parameter names.
+        num_bound_params = 1 if param_names[0] in ("self", "cls") else 0
+        func_name = func.__name__
+        if func_name == "__init__":
+            # In the warning, show "ClassName(...)" rather than "__init__(...)".
+            # The class isn't defined yet, but its name is in __qualname__.
+            # Some examples: "ClassName.__init__", "Nested.ClassName.__init__",
+            # "MyTests.test_case.<locals>.ClassName.__init__".
+            try:
+                local_name = func.__qualname__.rsplit("<locals>.", 1)[-1]
+                class_name = local_name.replace(".__init__", "")
+            except (AttributeError, IndexError, TypeError):
+                # __init__ outside a class or other oddity. The __name__ is fine.
+                pass
+            else:
+                func_name = class_name
+
+        if num_positional_params > num_bound_params:
+            # Clarify that only "some" (not all) positional arguments are affected.
+            message_template = (
+                "Use of some positional arguments is deprecated."
+                " Change to `{replacement}`."
+            )
+        else:
+            message_template = (
+                "Use of positional arguments is deprecated."
+                " Change to `{replacement}`."
```

Comment:
```suggestion
                "Passing positional arguments to {func_name}() is deprecated."
                " Change to {replacement}."
```
Been checking the previous deprecation messages and it's usually of the format `Passing positional arguments to {func_name}() is deprecated.`
I would remove the ` from the messages as we are not in the habit of including them
Note that the "Change to" is a nice addition but I wouldn't say necessary, I guess especially in the "everything should be a kwarg" case

Reply:
Specific suggestion [below](#discussion_r2167302580).